php/class-customize-concurrency.php

```diff
@@ -14,4 +14,299 @@
+
+		?>
+		<script type="text/html" id="tmpl-customize-post-section-notifications">
+			<ul>
+				<# _.each( data.notifications, function( notification ) { #>
+					<li class="notice notice-{{ notification.type || 'info' }} {{ data.altNotice ? 'notice-alt' : '' }}" data-code="{{ notification.code }}" data-type="{{ notification.type }}">
+						<# if ( /setting_update_conflict/.test( notification.code ) ) { #>
+							<button class="button override-post-conflict" type="button"><?php esc_html_e( 'Override', 'customize-posts' ); ?></button>
+							<# } #>
+								{{ notification.message || notification.code }}
+					</li>
+					<# } ); #>
+			</ul>
+		</script>
+		<?php
+	}
+
+
+	/**
+	 * Prevent sanitize_title() on the post_name.
+	 *
+	 * @see wp_insert_post()
+	 * @filter wp_insert_post_data
+	 *
+	 * @param array $data    An array of slashed post data.
+	 * @param array $postarr An array of sanitized, but otherwise unmodified post data.
+	 * @return array
+	 */
+	public function preserve_inserted_post_name( $data, $postarr ) {
+		if ( self::POST_TYPE === $data['post_type'] && isset( $postarr['post_name'] ) ) {
+			$data['post_name'] = wp_slash( $postarr['post_name'] );
+		}
+		return $data;
+	}
+
+	public function customize_save ( \WP_Customize_Manager $wp_customize ) {
+
+		$post_values = $wp_customize->unsanitized_post_values();
+		$saved_settings = $this->get_saved_settings( array_keys( $post_values ) );
+		$timestamps = isset( $_POST['concurrency_timestamps'] ) ? (array) json_decode( wp_unslash( $_POST['concurrency_timestamps'] ) ) : array();
+
+		$invalidities = array();
+
+		foreach ( $saved_settings as $setting_id => $saved_setting ) {
+			$is_conflicted = (
+				isset( $saved_setting[ 'timestamp' ], $saved_setting[ 'value' ] )
+				&&
+				$saved_setting['timestamp'] > $timestamps[ $setting_id ]
+				&&
+				$saved_setting['value'] !== $post_values[ $setting_id ]
+			);
+			if ( $is_conflicted ) {
+				$user = get_user_by( 'ID', (int) $saved_setting['author'] );
+				$message = sprintf( 'Value from %s: %s', $user->user_nicename, $saved_setting['value'] );
+				$invalidities[ $setting_id ] = new \WP_Error( 'concurrency_conflict', $message, array( 'their_value' => $saved_setting['value'] ) );
+			}
+		}
+
+		if ( ! empty( $invalidities ) ) {
+			$exported_setting_validities = array_map( array( $wp_customize, 'prepare_setting_validity_for_js' ), $invalidities );
+			$invalid_setting_count = count( $exported_setting_validities );
+			$response = array(
+				'setting_validities' => $exported_setting_validities,
+				'message' => sprintf(
+					_n( 'There is %s conflicting setting.', 'There are %s conflicting settings.', $invalid_setting_count ),
+					number_format_i18n( $invalid_setting_count )
+				),
+			);
+
+			/** This filter is documented in wp-includes/class-wp-customize-manager.php */
+			$response = apply_filters( 'customize_save_response', $response, $this );
+			wp_send_json_error( $response );
```

Comment:
@lgedeon I just realized this will be incompatible with Customize Snapshots when a snapshot is _scheduled_ or published via the WP Admin, outside the context of Customizer `customize_save` Ajax request. So we'll need to eliminate `wp_send_json_error()` here and instead go the alternative route we discussed of adding `customize_validate_{$setting_id}` filters, like so:

```php
add_action( 'customize_save_validation_before', function( \WP_Customize_Manager $wp_customize ) {

	$post_values = $wp_customize->unsanitized_post_values();
	$saved_settings = $this->get_saved_settings( array_keys( $post_values ) );

	// @todo When publishing a snapshot outside the Customizer (e.g. from admin or WP Cron), the setting $timestamps should get supplied from modified times for setting updates in the snapshot?
	$timestamps = isset( $_POST['concurrency_timestamps'] ) ? (array) json_decode( wp_unslash( $_POST['concurrency_timestamps'] ) ) : array();

	$validate_concurrency_conflict = function( \WP_Error $validity, $value, \WP_Customize_Setting $setting ) use ( $timestamps, $saved_settings ) {
		$is_conflicted = (
			array_key_exists( $setting->id, $saved_settings )
			&&
			isset( $saved_setting['timestamp'], $saved_setting['value'] )
			&&
			$saved_setting['timestamp'] > $timestamps[ $setting->id ]
			&&
			$saved_setting['value'] !== $value
		);
		if ( $is_conflicted ) {
			$user = get_user_by( 'ID', (int) $saved_setting['author'] );
			$message = sprintf( 'Value from %s: %s', $user->user_nicename, $saved_setting['value'] );
			$validity->add( 'concurrency_conflict', $message, array( 'their_value' => $saved_setting['value'] ) );
		}
		return $validity;
	};

	// @todo Adding a filter for validation will *FAIL* if a WP_Customize_Setting subclass overrides the validate() method and does not apply_filters( "customize_validate_{$setting->id}" );. A Core improvement is needed here to filter setting_validities to ensure plugins can do last minute validation of the values prior to save.
	foreach ( $wp_customize->settings() as $setting ) {
		add_filter( "customize_validate_{$setting->id}", $validate_concurrency_conflict, 1000, 3 );
	}
} );
```